staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go

```diff
@@ -164,37 +168,61 @@ func RunTestWatchFromZero(ctx context.Context, t *testing.T, store storage.Inter
 		t.Fatalf("GuaranteedUpdate failed: %v", err)
 	}
 
+	// If testing against the cacher, we allow for slow processing by allowing
+	// the updates to propagate to the watch cache.
+	if cacher {
+		time.Sleep(100 * time.Millisecond)
+	}
 	// Make sure when we watch from 0 we receive an ADDED event
 	w, err = store.Watch(ctx, key, storage.ListOptions{ResourceVersion: "0", Predicate: storage.Everything})
 	if err != nil {
 		t.Fatalf("Watch failed: %v", err)
 	}
 	testCheckResult(t, watch.Added, w, out)
-	w.Stop()
 
-	if compaction == nil {
-		t.Skip("compaction callback not provided")
-	}
-
-	// Update again
+	// Update existing object
 	out = &example.Pod{}
 	err = store.GuaranteedUpdate(ctx, key, out, true, nil, storage.SimpleUpdate(
 		func(runtime.Object) (runtime.Object, error) {
-			return &example.Pod{ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: "test-ns"}}, nil
+			return &example.Pod{ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: "test-ns", Annotations: map[string]string{"a": "2"}}}, nil
 		}), nil)
 	if err != nil {
 		t.Fatalf("GuaranteedUpdate failed: %v", err)
 	}
 
-	// Compact previous versions
-	compaction(ctx, t, out.ResourceVersion)
+	// If testing against the cacher, we allow for slow processing by allowing
+	// the updates to propagate to the watch cache.
+	if cacher {
+		time.Sleep(100 * time.Millisecond)
+	}
+	testCheckResult(t, watch.Modified, w, out)
+	w.Stop()
 
-	// Make sure we can still watch from 0 and receive an ADDED event
-	w, err = store.Watch(ctx, key, storage.ListOptions{ResourceVersion: "0", Predicate: storage.Everything})
-	if err != nil {
-		t.Fatalf("Watch failed: %v", err)
+	if !cacher {
```

Comment:
This if isn't needed, right?

We know that compaction isn't passed for cacher, so let's remove this if.

Reply:
Actually - thinking about it more - we can implement compaction for the test purposes in watchcache.

Effectively what we can do is:
1. create a test wrapper around Cacher that implements a separate compation method
2. the compaction method is implemented as:
```
take the lock on watchcache
pass the startPoint in the buffer to eliminate all events older than the given RV
adjust the listResourceVersion field accordingly
adjust removedEventSinceRelist field accordingly
[potentially] figure out of some watches should be closed
```

We definitely don't want that to be exposed anywhere, but for the purposes of tests it seems doable.

Reply:
@wojtek-t - do we still want to do this? I tried this out and came up with https://github.com/MadhavJivrajani/kubernetes/commit/5c60d4dbf077462e61c0955695619e5471e78cfb

I'm not sure if I am a fan since we are exposing a Cacher with this compaction method (verbosely names to indicate test only, but still). I would have preferred to have this isolated to the testing package but since `watchCache` is a private field in the `Cacher`, we can't really do the above manipulation. If we still want to do this for testing purposes and this approach seems okay, I can open a PR and we can iterate there

Reply:
I would still like this, but as you mentioned, it shouldn't be added into production code.

The way I was thinking about it is by:
- having a wrapper around cacher (defined in cacher package, but in a test file)
- passing some callback to the test itself
- [additionally finally moving the pkg/storage/tests to pkg/storage/cacher - similarly as we have for etcd]